src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/DefaultOutputFormatterSelector.cs

```diff
@@ -0,0 +1,366 @@
+                // are not currently respecting the 'browser accept header'.
+                _logger.NoAcceptForNegotiation();
+
+                selectFormatterWithoutRegardingAcceptHeader = true;
+            }
+            else
+            {
+                if (contentTypes.Count == 0)
+                {
+                    // Use whatever formatter can meet the client's request
+                    selectedFormatter = SelectFormatterUsingSortedAcceptHeaders(
+                        context,
+                        formatters,
+                        acceptableMediaTypes);
+                }
+                else
+                {
+                    // Verify that a content type from the context is compatible with the client's request
+                    selectedFormatter = SelectFormatterUsingSortedAcceptHeadersAndContentTypes(
+                        context,
+                        formatters,
+                        acceptableMediaTypes,
+                        contentTypes);
+                }
+
+                if (selectedFormatter == null && !_returnHttpNotAcceptable)
+                {
+                    _logger.NoFormatterFromNegotiation(acceptableMediaTypes);
+
+                    selectFormatterWithoutRegardingAcceptHeader = true;
+                }
+            }
+
+            if (selectFormatterWithoutRegardingAcceptHeader)
+            {
+                if (contentTypes.Count == 0)
+                {
+                    selectedFormatter = SelectFormatterNotUsingContentType(
+                        context,
+                        formatters);
+                }
+                else
+                {
+                    selectedFormatter = SelectFormatterUsingAnyAcceptableContentType(
+                        context,
+                        formatters,
+                        contentTypes);
+                }
+            }
+
+            if (selectedFormatter == null)
+            {
+                // No formatter supports this.
+                _logger.NoFormatter(context);
+                return null;
+            }
+
+            _logger.FormatterSelected(selectedFormatter, context);
+            return selectedFormatter;
+        }
+
+        private List<MediaTypeSegmentWithQuality> GetAcceptableMediaTypes(HttpRequest request)
+        {
+            var result = new List<MediaTypeSegmentWithQuality>();
+            AcceptHeaderParser.ParseAcceptHeader(request.Headers[HeaderNames.Accept], result);
+            for (var i = 0; i < result.Count; i++)
+            {
+                var mediaType = new MediaType(result[i].MediaType);
+                if (!_respectBrowserAcceptHeader && mediaType.MatchesAllSubTypes && mediaType.MatchesAllTypes)
+                {
+                    result.Clear();
+                    return result;
+                }
+            }
+
+            result.Sort((left, right) => left.Quality > right.Quality ? -1 : (left.Quality == right.Quality ? 0 : 1));
```

Comment:
Unrelated: should we cache this delegate?

Reply:
Sure - it will be cached already since it doesn't capture, but making it explicitly cached makes it harder to break